### PR TITLE
BAU: Add some security checks

### PR DESCRIPTION
**Pull request by tlwr**
- Fourth wall shouldn't be used over HTTP
- Fourth wall shouldn't use an access token which has unnecessary
permissions.

### CODE REVIEW

File: javascript/preflight.js [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+  const token             = new FourthWall.getQueryVariables().token,
+        ghUrl             = 'https://api.github.com/rate_limit',
+        authGhUrl         = ghUrl + '?access_token=' + token;
+
+  fetch(authGhUrl)
+    .then(function (response) { return response.headers; })
+    .then(function (headers)  { return headers.get('x-oauth-scopes')  })
+    .then(function (scopes)   { return scopes.split(', '); })
+    .then(function (scopes)   {
+      const allowedScopes = ['public_repo', 'repo:status', 'repo_deployment'];
```
Comment (davbo) [inline javascript/preflight.js]: even `public_repo` seems unnecessary for fourth wall ?

> Grants read/write access to code, commit statuses, collaborators, and deployment statuses for public repositories and organizations. Also required for starring public repositories.

Comment (tlwr) [reply javascript/preflight.js]: Good catch

Comment (davbo) [reply javascript/preflight.js]: @tlwr did you validate it works without `public_repo` ? I'm not a fourth wall user so my comment was just a guess.

Comment (tlwr) [reply javascript/preflight.js]: yes, only those two scopes are needed

File: javascript/preflight.js [hunk 2]
```diff
@@ -0,0 +1,46 @@
+function runPreFlightChecks() {
+  const isHttp      = window.location.protocol == 'http:',
+        isLocalhost = window.location.hostname == 'localhost'
+        isUnsafe    = !(isHttp && isLocalhost);
+
+  if (isUnsafe) {
+    const isHttpMessage = [
+      'This page is running over the web over HTTP.',
+      'You should use HTTP and rotate your access token.',
```
Comment (issyl0) [inline javascript/preflight.js]: Shouldn't this be "You should use HTTPS"?

Comment (tlwr) [reply javascript/preflight.js]: Correct, amended.

File: javascript/preflight.js [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+  const token             = new FourthWall.getQueryVariables().token,
+        ghUrl             = 'https://api.github.com/rate_limit',
+        authGhUrl         = ghUrl + '?access_token=' + token;
+
+  fetch(authGhUrl)
+    .then(function (response) { return response.headers; })
+    .then(function (headers)  { return headers.get('x-oauth-scopes')  })
+    .then(function (scopes)   { return scopes.split(', '); })
+    .then(function (scopes)   {
+      const allowedScopes = ['repo:status', 'repo_deployment'];
```
Comment (davbo) [inline javascript/preflight.js]: We need to add `read:org` to this list as it's used to get repos by team membership (mentioned in the README):
https://github.com/alphagov/fourth-wall#how-to-use

I'd also suggest updating the README to be more clear about which scopes are necessary and which (read:org) are optional. 

After that I'd be happy to merge.

**Comment by tlwr** [comment]
Updated this with above @davbo

File: javascript/preflight.js [hunk 4]
```diff
@@ -0,0 +1,46 @@
+function runPreFlightChecks() {
+  const isHttp      = window.location.protocol == 'http:',
+        isLocalhost = window.location.hostname == 'localhost'
+        isUnsafe    = !(isHttp && isLocalhost);
```
Comment (edds) [inline javascript/preflight.js]: This seems to have broken fourth-wall for us. We access it using the url: `https://alphagov.github.io/fourth-wall/..`

So this code reads as:
```javascript
const isHttp      = window.location.protocol == 'http:',  // https: == 'http:' => false
      isLocalhost = window.location.hostname == 'localhost' // alphagov.github.io == 'localhost' => false
      isUnsafe    = !(isHttp && isLocalhost); // '!(false && false) => !false => true
```

So even though we are using it over https this is triggering as unsafe. I think the logic you want is:

```javascript
isUnsafe = !http || !(http && isLocalhost)
```
which means if your on https you're safe, or if your on http and localhost your safe.

**Comment by edds** [comment]
This change blocks us from giving fourth wall access to read information about private repos, as it thinks we don't need permissions it needs to be able to do that, and stops us from using it at the `https://alphagov.github.io` url.

**Pull request merged**